pkg/relayer/proxy/websockets/bridge.go

```diff
@@ -0,0 +1,379 @@
+
+	logger.Debug().Msg("relay emitted to miner")
+
+	// Accumulate the relay reward.
+	if err := b.relayMeter.AccumulateRelayReward(b.ctx, relayRequest.Meta); err != nil {
+		b.serviceBackendConn.handleError(
+			ErrWebsocketsGatewayMessage.Wrapf("failed to accumulate relay reward: %v", err),
+		)
+		return
+	}
+
+}
+
+func (b *bridge) handleServiceBackendIncomingMessage(msg message) {
+	logger := b.logger.With(
+		"message_source", messageSourceServiceBackend,
+		"message_type", msg.messageType,
+	)
+
+	logger.Debug().Msg("received message from service backend")
+
+	// Use the latest relay request's session header to create the RelayResponse.
+	meta := b.latestRelayRequest.Meta
+	serviceId := meta.SessionHeader.ServiceId
+
+	// Create a RelayResponse from the service backend message.
+	relayResponse := &types.RelayResponse{
+		Meta:    types.RelayResponseMetadata{SessionHeader: meta.SessionHeader},
+		Payload: msg.data,
+	}
+
+	relayer.RelaysTotal.With(
+		"service_id", serviceId,
+		"supplier_operator_address", meta.SupplierOperatorAddress,
+	).Add(1)
+
+	relayer.RelayResponseSizeBytes.With("service_id", serviceId).
+		Observe(float64(relayResponse.Size()))
+
+	// Sign the relay response and add the signature to the relay response metadata
+	if err := b.relayAuthenticator.SignRelayResponse(relayResponse, meta.SupplierOperatorAddress); err != nil {
+		b.gatewayConn.handleError(
+			ErrWebsocketsServiceBackendMessage.Wrapf("failed to sign relay response: %v", err),
+		)
+		return
+	}
+
+	logger.Debug().Msg("relay response signed")
+
+	// Store the latest relay response to guarantee that there is always a request/response.
+	b.setLatestRelayResponse(relayResponse)
+
+	relayResponseBz, err := relayResponse.Marshal()
+	if err != nil {
+		b.gatewayConn.handleError(
+			ErrWebsocketsServiceBackendMessage.Wrapf("failed to marshal relay response: %v", err),
+		)
+		return
+	}
+
+	// Forward the relay response to the gateway.
+	if err := b.gatewayConn.WriteMessage(msg.messageType, relayResponseBz); err != nil {
+		b.gatewayConn.handleError(
+			ErrWebsocketsServiceBackendMessage.Wrapf("failed to send relay response to gateway: %v", err),
+		)
+		return
+	}
+
+	logger.Debug().Msg("relay response forwarded to gateway")
+
+	// Do not emit a relay to the miner if there is no response to form a request/response pair.
+	if b.getLatestRelayRequest() == nil {
+		logger.Info().Msg("waiting for service backend request")
+		return
+	}
+
+	relay := &types.Relay{
+		Req: b.getLatestRelayRequest(),
+		Res: relayResponse,
+	}
+
+	relayer.RelaysSuccessTotal.With("service_id", serviceId).Add(1)
+
+	// Emit the relay to the miner.
+	// Since async relays might be request or response only, each request or response
+	// is considered to be eligible for a reward.
+	b.relaysProducer <- relay
+
+	logger.Debug().Msg("relay emitted to miner")
+
+	// Accumulate the relay reward.
+	if err := b.relayMeter.AccumulateRelayReward(b.ctx, b.latestRelayRequest.Meta); err != nil {
```

Comment:
Per my comment above, we're accumulating relay rewards in both directions, is that right?

Feels like websockets are "easy" because it's just HTTP over a stream that doesn't change.

If (when?) we go with audio/video streams, we'd need to do something with delimiters and data sizes.

My intention here isn't to add a TODO, but just ensure I'm not missing anything. Confirmed?

Reply:
You are :100: correct.

Will add the following:
```
TODO_FUTURE: Currently, the RelayMiner is payed for each incoming and outgoing message transmitted.
While this is the most common and trivial use case, future services might have different payable units of work
(e.g. packet size, specific packet or data delimiter...). 
```